src/udev/udev-builtin-net_id.c

```diff
@@ -600,6 +603,101 @@ static int names_platform(sd_device *dev, NetNames *names, bool test) {
         return 0;
 }
 
+#define DEVICETREE_ALIASES "/proc/device-tree/aliases"
+
+static int dev_devicetree_onboard(sd_device *dev, NetNames *names) {
+        char *ofnode;
+        const char *parent_syspath;
+        _cleanup_free_ char *devicetree_syspath = NULL;
+        _cleanup_free_ char *ofnode_syspath = NULL;
+        _cleanup_free_ char *ofnode_path = NULL;
+        _cleanup_closedir_ DIR *aliases = NULL;
+        sd_device *parent = NULL;
+        int r;
+        size_t ofnode_pathlen;
+        unsigned int index = 0;
+
+        if (!naming_scheme_has(NAMING_DEVICETREE_ALIASES))
+                return 0;
+
+        /* check if our direct parent has an of_node */
+        r = sd_device_get_parent(dev, &parent);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_syspath(parent, &parent_syspath);
+        if (r < 0)
+                return r;
+
+        ofnode = strjoina(parent_syspath, "/of_node");
+        r = chase_symlinks(ofnode, NULL, 0, &ofnode_syspath, NULL);
+        if (r < 0)
+                return r;
+
+        r = chase_symlinks("/proc/device-tree", NULL, 0, &devicetree_syspath, NULL);
+        if (r < 0)
+                return r;
+
+        /*
+         * Example paths:
+         * ofnode_syspath = /sys/firmware/devicetree/base/soc/ethernet@deadbeef
+         * devicetree_syspath = /sys/firmware/devicetree/base
+         * ofnode_path = soc/ethernet@deadbeef
+         */
+        r = path_make_relative(devicetree_syspath, ofnode_syspath, &ofnode_path);
+        if (r < 0)
+                return r;
+
+        if (startswith(ofnode_path, ".."))
+                /* the of_node is from another devicetree */
+                return -ENOENT;
+        ofnode_pathlen = strlen(ofnode_path);
+
+        aliases = opendir(DEVICETREE_ALIASES);
+        if (!aliases)
+                return -errno;
+
+        FOREACH_DIRENT_ALL(alias, aliases, break) {
+                const char *conflict = DEVICETREE_ALIASES "/ethernet";
+                _cleanup_free_ char *alias_path = NULL;
+                char alias_syspath[PATH_MAX];
+                size_t alias_pathlen;
+
+                if (!startswith(alias->d_name, "ethernet") ||
+                    !snprintf_ok(alias_syspath, sizeof(alias_syspath),
+                                 DEVICETREE_ALIASES "/%s", alias->d_name) ||
+                    !read_full_virtual_file(alias_syspath, &alias_path, &alias_pathlen))
```

Comment:
You can use `read_virtual_file_at()`, then it is not necessary to build `alias_syspath`.

Also, `read_full_virtual_file(...)` returns negative errno on error, hence please use `< 0`.

Reply:
OK

Reply:
As it turns out, I can use sd_device_new_from_syspath on /aliases, which lets me use sd_device_get_sysattr_value to get the aliases instead of doing the reads directly.